x/gamm/keeper/pool_twap.go

```diff
@@ -0,0 +1,234 @@
+	pool, err := k.GetPool(ctx, poolId)
+	if err != nil {
+		return types.PoolTwapHistory{}, err
+	}
+
+	var twapPairs []*types.TwapPair
+	// iterate through all assets, creating all possible pairs
+	for i, tokenIn := range pool.GetAllPoolAssets() {
+		for j, tokenOut := range pool.GetAllPoolAssets() {
+			// if it is not the same token, create a twap pair
+			if i != j {
+				spotPrice, err := k.CalculateSpotPrice(ctx, poolId, tokenIn.Token.Denom, tokenOut.Token.Denom)
+				if err != nil {
+					return types.PoolTwapHistory{}, err
+				}
+				twapPair := types.TwapPair{
+					TokenIn:         tokenIn.Token.Denom,
+					TokenOut:        tokenOut.Token.Denom,
+					PriceCumulative: sdk.ZeroDec(),
+					SpotPrice:       spotPrice,
+				}
+				twapPairs = append(twapPairs, &twapPair)
+			}
+		}
+	}
+
+	poolTwap := types.PoolTwapHistory{
+		TimeStamp: ctx.BlockTime(),
+		PoolId:    poolId,
+		TwapPairs: twapPairs,
+	}
+	return poolTwap, nil
+}
+
+// update pool twap with token(s) that have changed
+func (k Keeper) UpdatePoolTwap(ctx sdk.Context, poolId uint64, tokens ...string) (err error) {
+	currentTime := ctx.BlockTime()
+	recentPoolTwap, err := k.GetPoolTwapHistory(ctx, poolId, currentTime)
+	if err != nil {
+		return err
+	}
+
+	if len(tokens) > 3 {
+		return fmt.Errorf("tokens should be two or less")
+	}
+
+	// iterate through the array of spot prices,
+	// updating all spot prices that are related to the changed token
+	for i, spotPrice := range recentPoolTwap.TwapPairs {
+		currentTimeElapsedDuration := currentTime.Sub(recentPoolTwap.TimeStamp)
+
+		// division between integer, since minimum unit for duration is seconds
+		currentTimeElapsed := currentTimeElapsedDuration.Nanoseconds() / 1_000_000_000
+
+		// if any of the spot price pairs in twap history are realted to the swapped tokens,
+		// update the spot price cumulative
+		if contains(tokens, spotPrice.TokenIn, spotPrice.TokenOut) {
+			changedSpotPrice, err := k.CalculateSpotPrice(ctx, poolId, spotPrice.TokenIn, spotPrice.TokenOut)
+			if err != nil {
+				return err
+			}
+			// recentPoolTwap.TwapPairs[i].PriceCumulative = recentPoolTwap.TwapPairs[i].PriceCumulative.Add(changedSpotPrice.Mul(currentTimeElapsed))
+			recentPoolTwap.TwapPairs[i].PriceCumulative = recentPoolTwap.TwapPairs[i].PriceCumulative.Add(changedSpotPrice.Mul(sdk.NewDec(currentTimeElapsed)))
+			recentPoolTwap.TwapPairs[i].SpotPrice = changedSpotPrice
+		} else {
+			recentPoolTwap.TwapPairs[i].PriceCumulative = recentPoolTwap.TwapPairs[i].PriceCumulative.Add(recentPoolTwap.TwapPairs[i].SpotPrice.Mul(sdk.NewDec(currentTimeElapsed)))
+		}
+	}
+
+	poolTwap := types.PoolTwapHistory{
+		TimeStamp: ctx.BlockTime(),
+		PoolId:    poolId,
+		TwapPairs: recentPoolTwap.TwapPairs,
+	}
+
+	k.SetPoolTwap(ctx, poolTwap)
+	return nil
+}
+
+func (k Keeper) GetPoolTwapKey(ctx sdk.Context, poolId uint64, queryTime time.Time) []byte {
+	store := ctx.KVStore(k.storeKey)
+	timestamp := queryTime.Unix()
```

Comment:
(Note to Self): TODO, double check this is what we normally do for converting timestamps to store keys across the SDK